core/string/ustring.cpp

```diff
@@ -3072,15 +3072,24 @@ int String::rfindn(const String &p_str, int p_from) const {
 
 bool String::ends_with(const String &p_string) const {
 	int l = p_string.length();
+	if (l > length()) {
+		return false;
+	}
+
 	if (l == 0) {
 		return true;
 	}
 
-	int pos = rfind(p_string);
-	if (pos == -1) {
-		return false;
+	const char32_t *p = &p_string[0];
+	const char32_t *s = &operator[](length() - l);
+
+	for (int i = 0; i < l; i++) {
+		if (p[i] != s[i]) {
+			return false;
+		}
 	}
-	return pos + l == length();
+
+	return true;
```

Comment:
Would be worth refactoring the `begins_with` code to be the same, there's no reason to have differences in those two nearly identical methods (initial `l` check, return `true` vs unnecessary test for `i == l`).

Reply:
Right. I've changed `begin_with` taking in a `String` as a parameter to be analogous.